streamalert/artifact_extractor/artifact_extractor.py

```diff
@@ -0,0 +1,238 @@
+
+        Kwargs (optional):
+            function (str): Describes how this field is used in the record, or what it means.
+            record_id (str): Currently it is reserved for future support. It will come from the
+                record processed by classifier. This field is very useful for cross reference back
+                to the original record in the future.
+        """
+        self._function = kwargs.get('function', 'not_specified')
+        self._record_id = kwargs.get('record_id', 'RESERVED')
+        self._source_type = source_type
+        self._type = normalized_type
+        self._value = value
+
+    @property
+    def record(self):
+        """Generate an artifact
+
+        Returns:
+            dict: A dictionary contains artifact information.
+        """
+        return {
+            'function': self._function,
+            'record_id': self._record_id,
+            'source_type': self._source_type,
+            'type': self._type,
+            'value': self._value,
+        }
+
+
+class FirehoseRecord:
+    """Encapsulation of single Firehose record and/or normalized artifacts"""
+
+    def __init__(self, firehose_record, source_type):
+        """Create a new Firehose record contains original data and may extract multiple artifacts if
+        original data was normalized in the classifier.
+        The original data which will be returned back to source firehose for historical search. And
+        the artifacts, if any, will be sent to a dedicated firehose with simplified schema and land
+        in streamalert data bucket for historical search as well.
+
+        Args:
+            firehose_record (dict): the record passed to lambda from source firehose. It has
+                following format,
+
+                {
+                  'recordId': '12345678901230000000',
+                  'data': 'eyJyeXhpYXMiOiJZZXMiLCJtZXNzYWdlIjoiaGVsbG8gd29ybGQhIiwiZXZlbnRfZG==',
+                  'approximateArrivalTimestamp': 1583275634682
+                }
+
+            source_type (str): The original source of the artifact(s) extracted from a record.
+                e.g. osquery_differential, cloudwatch_cloudtrail
+        """
+        self._firehose_record_id = firehose_record['recordId']
+        self._firehose_data = firehose_record['data']
+        self._decoded_record = json.loads(base64.b64decode(self._firehose_data))
+        self._source_type = source_type
+
+    @property
+    def artifacts(self):
+        """Extract all artifacts from a record
+
+        Returns:
+            list: A list of Artifacts from a normalized record.
+        """
+        artifacts = []
+
+        if Normalizer.NORMALIZATION_KEY not in self._decoded_record:
+            # Return an empty list if the record doesn't have normalization information.
+            return artifacts
+
+        if not self._source_type:
+            # Return immediately if can not identify source_type. a.k.a do not extract artifacts.
```

Comment:
Maybe log a warning?

Reply:
👍 It generates a warning message in `def _get_source_type(self):` method when parse the source type from the first place.